datalad/plugin/addurls.py

```diff
@@ -569,13 +569,19 @@ def add_meta(rows):
     """
     for row in rows:
         ds, filename = row["ds"], row["ds_filename"]
-        lgr.debug("Adding metadata to %s in %s", filename, ds.path)
-        for a in ds.repo.set_metadata(filename, add=row["meta_args"]):
-            res = annexjson2result(a, ds, type="file", logger=lgr)
-            # Don't show all added metadata for the file because that
-            # could quickly flood the output.
-            del res["message"]
-            yield res
+
+        old_always_commit = ds.repo.always_commit
+        ds.repo.always_commit = False
```

Comment:
FWIW mock is our core dependency, not only for the tests, so you could use its patch here to minimize boiler plate, 

But isn't there a need to make the last call to set_metadata without that config flag to cause staged metadata changes to be actually committed (in --no-save mode, or is that oxymoronic??? ;))

Reply:
> FWIW mock is our core dependency, not only for the tests, so you could use its patch here to minimize boiler plate,

Ah, hadn't thought about using mock here.  I had a commit that introduced a `no_annex_commits` context manager to utils.py, but it couldn't be easily used in the one other place in the code base that does this `always_commit` set/reset dance, so I just dropped it.

I'll switch to using mock here.

> But isn't there a need to make the last call to set_metadata without that config flag to cause staged metadata changes to be actually committed (in --no-save mode, or is that oxymoronic??? ;))

Right, the commit message touches on this: "If --nosave is given, the metadata changes will remain in .git/annex/journal until an action outside of addurls triggers the commit".  I went back and forth on whether we should make `add_meta` trigger the commit.  I suppose the main worry is that the changes could be dropped from the journal (perhaps if `git gc` is run before another action triggers the commit?  I haven't tested).

If we want to make `meta` trigger the commit, we could do that by adjusting the last call like you mentioned, but IMO it's more straightforward to trigger the commit by making a separate noop git-annex call in each dataset.  I have a patch locally that does that.